In [17]:
# %matplotlib widget

from __future__ import annotations

import re
from collections import defaultdict
from dataclasses import dataclass, field
from itertools import permutations, product
from math import inf
from random import choice

import matplotlib.colors as mcolors
import matplotlib.pyplot as plt
import networkx as nx
import numpy as np
import numpy.typing as npt
from mpl_toolkits.mplot3d import axes3d
from numpy import int_, object_
from numpy.typing import NDArray
from test_utilities import run_tests_params
from util import print_hex

COLORS = list(mcolors.CSS4_COLORS.keys())

<link href="style.css" rel="stylesheet"></link>
<article class="day-desc read-aloud"><h2>--- Day 5: A Maze of Twisty Trampolines, All Alike ---</h2><p>An urgent <span title="Later, on its turn, it sends you a sorcery.">interrupt</span> arrives from the CPU: it's trapped in a maze of jump instructions, and it would like assistance from any programs with spare cycles to help find the exit.</p>
<p>The message includes a list of the offsets for each jump. Jumps are relative: <code>-1</code> moves to the previous instruction, and <code>2</code> skips the next one. Start at the first instruction in the list. The goal is to follow the jumps until one leads <em>outside</em> the list.</p>
<p>In addition, these instructions are a little strange; after each jump, the offset of that instruction increases by <code>1</code>. So, if you come across an offset of <code>3</code>, you would move three instructions forward, but change it to a <code>4</code> for the next time it is encountered.</p>
<p>For example, consider the following list of jump offsets:</p>
<pre><code>0
3
0
1
-3
</code></pre>
<p>Positive jumps ("forward") move downward; negative jumps move upward. For legibility in this example, these offset values will be written all on one line, with the current instruction marked in parentheses. The following steps would be taken before an exit is found:</p>
<ul>
<li><code>(0)&nbsp;3&nbsp;&nbsp;0&nbsp;&nbsp;1&nbsp;&nbsp;-3&nbsp;</code> - <em>before</em> we have taken any steps.</li>
<li><code>(1)&nbsp;3&nbsp;&nbsp;0&nbsp;&nbsp;1&nbsp;&nbsp;-3&nbsp;</code> - jump with offset <code>0</code> (that is, don't jump at all). Fortunately, the instruction is then incremented to <code>1</code>.</li>
<li><code>&nbsp;2&nbsp;(3)&nbsp;0&nbsp;&nbsp;1&nbsp;&nbsp;-3&nbsp;</code> - step forward because of the instruction we just modified. The first instruction is incremented again, now to <code>2</code>.</li>
<li><code>&nbsp;2&nbsp;&nbsp;4&nbsp;&nbsp;0&nbsp;&nbsp;1&nbsp;(-3)</code> - jump all the way to the end; leave a <code>4</code> behind.</li>
<li><code>&nbsp;2&nbsp;(4)&nbsp;0&nbsp;&nbsp;1&nbsp;&nbsp;-2&nbsp;</code> - go back to where we just were; increment <code>-3</code> to <code>-2</code>.</li>
<li><code>&nbsp;2&nbsp;&nbsp;5&nbsp;&nbsp;0&nbsp;&nbsp;1&nbsp;&nbsp;-2&nbsp;</code> - jump <code>4</code> steps forward, escaping the maze.</li>
</ul>
<p>In this example, the exit is reached in <code>5</code> steps.</p>
<p><em>How many steps</em> does it take to reach the exit?</p>
</article>


In [18]:
example = """
0
3
0
1
-3
"""


def steps_to_exiit(jumps: str) -> int:
    js = [int(i) for i in jumps.strip().splitlines()]
    counter, steps = 0, 0

    while 0 <= counter < len(js):
        jump = js[counter]
        js[counter] += 1
        counter += jump
        steps += 1

    return steps


steps_to_exiit(example)

5

In [19]:
with open("../input//day5.txt") as f:
    puzzle = f.read()

print(f"Part I: {steps_to_exiit(puzzle)}")

Part I: 364539


<link href="style.css" rel="stylesheet"></link>
<main>

<p>Your puzzle answer was <code>364539</code>.</p><p class="day-success">The first half of this puzzle is complete! It provides one gold star: *</p>
<article class="day-desc"><h2 id="part2">--- Part Two ---</h2><p>Now, the jumps are even stranger: after each jump, if the offset was <em>three or more</em>, instead <em>decrease</em> it by <code>1</code>. Otherwise, increase it by <code>1</code> as before.</p>
<p>Using this rule with the above example, the process now takes <code>10</code> steps, and the offset values after finding the exit are left as <code>2 3 2 3 -1</code>.</p>
<p><em>How many steps</em> does it now take to reach the exit?</p>
</article>

</main>


In [20]:
def stranger_steps_to_exit(jumps: str) -> int:
    js = [int(i) for i in jumps.strip().splitlines()]
    counter, steps = 0, 0

    while 0 <= counter < len(js):
        jump = js[counter]
        if js[counter] >= 3:
            js[counter] -= 1
        else:
            js[counter] += 1
        counter += jump
        steps += 1

    return steps


stranger_steps_to_exit(example)

10

In [21]:
print(f"Part II: {stranger_steps_to_exit(puzzle)}")

Part II: 27477714


<link href="style.css" rel="stylesheet"></link>
<main>

<p>Your puzzle answer was <code>27477714</code>.</p><p class="day-success">Both parts of this puzzle are complete! They provide two gold stars: **</p>

</main>
